GitHub actions

bdevierno1 opened the pull request:
Implement Github Actions

<!-- Add detailed description and provide running instructions -->
Automates part of our CI process with Github Actions. It checks if users have created a pr to the correct branch, performs linter on modified code. It lints files with all of the following: gwclint.py, shellcheck, cppcheck, pylint. More detailed explanations can be found in the style/github-action-style.md

**Usage:**

<!-- Check list of the things this PR accomplishes -->
| This PR includes         |          |
| ------------------------ | -------- |
| Resolves issues          | <!-- Provide a list of issues --> 
| Breaking API changes     |
| Internal API changes     |
| Usability improvements   |  
| Bug fixes                |
| New functionality        | 👍 
| New NF/onvm_mgr args     | 
| Changes to starting NFs  |  
| Dependency updates       | 
| Web stats updates        | 


<!-- If the pr has any dependencies or merge quirks note them here -->
## Merging notes:
 - Dependencies: None  

**TODO before merging :**
 - [x] PR is ready for review


<!-- What you did to test the PR, what needs to be done -->
## Test Plan:
* The Github Action is triggered by the get-user-lint.sh script. This script can be tested in your terminal. Detailed explanations of how to do so can be found in the style/github-action-style.md

<!-- Notes about what you think should be reviewed, any specific hacks in this PR -->
## Review: 
* There are different types of events that can be started by a Github action process. This workflow is kicked off when a pull request is made. It is subsequently restarted when a user makes a push to an open PR.

Special thanks to co-author @kevindweb who this would not be possible without.

onvm commented:
> In response to PR creation

## CI Message

Your results will arrive shortly

kevindweb commented:
great job! 🔥 Code looks clean, much better than the current CI linter process! Once this is merged, I'll remove the branch check and linter from CI. We probably won't have much use for unatuhorized CI runs after GitHub checks are created but that's ok!

kevindweb commented:
Just gave you access to run CI @bdevierno1 . Any ideas of how we can test this PR Ben? Could you provide a link to your test cases in your repository as well?

bdevierno1 commented:
> Just gave you access to run CI @bdevierno1 . Any ideas of how we can test this PR Ben? Could you provide a link to your test cases in your repository as well?

If you would like to test this within your own repository you would have to just copy over the two files and update line 12 of get-user-lint.sh to your own repo name. Create a pr from a new branch to your master/develop. If you click on Actions you will see a workflow would have started. 

I have three tests below I performed in my own repo.
* Two files updated with styling errors. New file with styling errors. Incorrect branch. https://github.com/bdevierno1/openNetVM/runs/519096130?check_suite_focus=true

* File updated with no styling issues.
 https://github.com/bdevierno1/openNetVM/runs/520348316?check_suite_focus=true

* File updated with no issues. Correct styling and correct branch.
https://github.com/bdevierno1/openNetVM/runs/520348316?check_suite_focus=true

You can view these three pull request here.
https://github.com/bdevierno1/openNetVM/pulls

kevindweb commented:
This does not seem promising... 
https://github.com/bdevierno1/openNetVM/pull/15/checks?check_run_id=520348316

Check out the last few lines of `User's Lint`. This is not a good sign I don't think
```
grep: file2.txt: No such file or directory
```
Seems like that's the only reason it passed because when the grep happens, it doesn't put anything into `diff.txt`. It doesn't seem like something we want in the Github check success.

bdevierno1 commented:
> https://github.com/bdevierno1/openNetVM/pull/15/checks?check_run_id=520348316

@kevindweb Thanks 👍 The issue was when new files were created, the file was could not be lint by the master. So if u made a pr with only new files, this would result in that issue as only stderr will write to file2. I added a line that has resolved this. Re-did all the tests and my previous comment has been updated with the new links.

kevindweb commented on style/get-user-lint.sh:
```diff
@@ -0,0 +1,44 @@
+done < "$input"
+git stash pop
+
+# Only output errors present in the user's changes and not in the master's lint. Save output to diff.txt .
```
```suggestion
# Only output errors present in the user's changes and not in the master's lint. Save output to diff.txt.
 ```

kevindweb commented on style/get-user-lint.sh:
```diff
@@ -0,0 +1,44 @@
+#! /bin/bash
+
+# A script to run linter functions on updated files. 
```
```suggestion
# A script to run linter functions on updated files.
```

kevindweb commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,35 @@
+      # Clone Repository
+      - name: Checkout
+        uses: actions/checkout@v1
+        
```
```suggestion

```

kevindweb replied:
These are not getting rid of the line, just the whitespace.

kevindweb commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,35 @@
+  lint:
+    runs-on: ubuntu-latest
+    steps: 
+      
```
```suggestion

```

kevindweb commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,35 @@
+          
+  lint:
+    runs-on: ubuntu-latest
+    steps: 
```
```suggestion
    steps:
```

kevindweb commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,35 @@
+        run: |
+          echo Run on develop not ${{github.base_ref}}
+          exit 1
+          
```
```suggestion

```

kevindweb commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,35 @@
+  branch: 
+    runs-on: ubuntu-latest
+    
+    steps: 
```
```suggestion
  branch:
    runs-on: ubuntu-latest
    steps:
```

kevindweb commented on style/get-user-lint.sh:
```diff
@@ -0,0 +1,44 @@
+#! /bin/bash
```
```suggestion
#!/bin/bash
```

kevindweb commented on style/get-user-lint.sh:
```diff
@@ -0,0 +1,44 @@
+
+# Check if the diff.txt file is empty. If it's not empty, errors are present and the lint exits 1.
+if [[ -s $diff_output ]]; then
+    echo $'\nYour lint results below. \n'
```
```suggestion
    echo $'\nYour lint results below.\n'
```

kevindweb commented:
Yay! People can look at Ben's forked repo if they want to see his testing. I think the code looks good. Ben and I have talked a lot about the updates he just pushed. The code pushed in the `get-user-lint.py` file is an updated version of the current linting that will be removed from CI.

Ben did a good job abstracting out the code for each of the different linters (bash, python, c, etc). Because of this, in the future, we will be able to add any new linter or tool with significant ease.

bdevierno1 commented:
> Made a bunch of nit suggestions. Here is what I use in vim to see the whitespace
> 
> ```viml
> :highlight ExtraWhitespace ctermbg=red guibg=red
> :match ExtraWhitespace /\s\+$/
> ```
> 
> I also feel that we should change the name from `get-user-lint.sh` to something more specific to C code, like `get-clint.sh` (maybe something more clever). As we add `*.sh` and `*.py` linter actions, we'll probably put those scripts in the `/style` directory as well. Keeping them unambiguous will be helpful.

Considering we have now added four different  types of static analysis, get-clint is now artifact. Please let me know if anyone has any suggestions for the nomenclature.

pcodes commented on style/get-user-lint.sh:
```diff
@@ -0,0 +1,56 @@
+# A script to run linter functions on updated C/python/bash files.
+# This script is triggered by github actions when pull request are created.
+# Users may also run this script to lint their file changes before pushing their code.
+# Usage: run ./style/get-c-lint.sh python OR ./style/get-c-lint.sh c OR ./style/get-c-lint.sh shell OR ./style/get-c-lint.sh cppcheck
```
Update the usage to reflect new file name

pcodes commented on style/get-user-lint.sh:
```diff
@@ -0,0 +1,56 @@
+
+# Check if user is in main git repository and correct number of arguments are passed in.
+if [ ! -d .git ]; then
+    echo 'Please run in main git repository. Example Usage: ./style/get-c-lint.sh python';
```
Update name

bdevierno1 commented on .github/workflows/Linter.yml:
```diff
@@ -0,0 +1,82 @@
+    steps:
+      # Clone Repository
+      - name: Checkout
+        uses: actions/checkout@v1
```
Thank you for your suggestions. I have updated it to v2 and tested it again. You can view the results of those tests here. https://github.com/bdevierno1/openNetVM/pull/47

twood02 commented:
@bdevierno1 -  Check if we can easily adjust the git diff line to make it include changes from uncommitted files (possibly just remove `...HEAD`)

bdevierno1 commented:
Since last review installation script has been added, updated such that users can test w/o having to commit. **Note for new files users still have to git add filename.** I have also updated how shellcheck parses the lint results in run-lint.py. 
Semi related to https://github.com/sdnfv/openNetVM/issues/161. Pylint will throw syntax errors for python2 scripts. This is especially noticeable for files with print statements w/o parantheses. Since we are using pylint-2.4 which relies on python3, it will exit once it reaches this line and not parse the rest of the file.

kevindweb commented on style/github-action-style.md:
```diff
@@ -0,0 +1,87 @@
+## Github Actions
+
+The **.github/workflows/Linter.yml** script is triggered when pull request are opened. The workflow is automatically re-run when a pull request is made. \
+The workflow has five seperate jobs that run concurrently.
```
```suggestion
The workflow has five separate jobs that run concurrently.
```

kevindweb commented on style/github-action-style.md:
```diff
@@ -0,0 +1,87 @@
+
+Run **./style/install-linters.sh** to install the different static tools.
+
+If you encounter errors during installation, you *may* install the latest version that yout OS supports with **pip3 install pylint** and **sudo apt-get install cppcheck**
```
```suggestion
If you encounter errors during installation, you *may* install the latest version that your OS supports with **pip3 install pylint** and **sudo apt-get install cppcheck**
```

kevindweb commented on style/install-linters.sh:
```diff
@@ -0,0 +1,20 @@
+pip3 --version
+pip3 install pylint==2.4.4
+export PATH=$HOME/.local/bin:$PATH
+pylint --version
```
```suggestion
pylint --version
echo export PATH="$HOME"/.local/bin:"$PATH" >> ~/.bashrc
echo "Please run source ~/.bashrc"
```
This is because exporting in a bash script does not export in the actual shell. We need the user to export it so they can run `pylint` on their own. We do this in the `install.sh` of onvm.

kevindweb commented:
@onvm final test

onvm commented:
> @onvm final test

## CI Message

Your results will arrive shortly
